coverage.out

```diff
@@ -0,0 +1,548 @@
+mode: set
```

Comment:
this file was generated from go test --cover-profile to get the coverage.out which is required for gocover.

Reply:
can we gitignore this?

Reply:
I think we need to generate this file on the fly within the pipeline, otherwise we will have to update it constantly

Reply:
How will this work when `main` changes? Is there something automated that will update this in `main`?

Reply:
ah my comment is the same as brandon's

Reply:
Actually if we need separation then coverage.out needs to be in the repo and added to .gitignore. if not we can combine them together since go test runs before go cover which would generate coverage.out and can be used by gocover. so we need to decided which approach to follow. Please let me know your thoughts.

Reply:
We should be generating the `coverage.out` by running go test on the pr 100%.

The big question is how we should be comparing to main. `main` branch needs its own `coverage.out` that we either automatically generate on the main branch or we attach to a workflow run and pull from the workflow run artifacts. I don't really like either of those solutions and would rather use Coveralls like agentbaker https://github.com/Azure/AgentBaker/blob/master/.github/workflows/check-coverage.yml.

Reply:
coverall requires workflow to have on push event, I don't know if it would work with repository dispatch_event like /ok-to-test. Please see screenshot below from coverall repo. 

![image](https://github.com/Azure/aks-app-routing-operator/assets/105889062/b7ca5588-4947-4bdb-9ec6-8030b54b3bac)

Yes totally agree we can generate coverage.out but separation can't be achieved. Or we can duplicate go test with cover profile again in go cover stage/job.

Reply:
There's no reason that unit tests can't be in a separate workflow than e2e tests. E2e is what needs repository_dispatch because that's the one that requires secrets.

Reply:
On the separation thing this would be part of the unit test workflow. You're right that there's no way to separate them.

It has to be part of the unit test job, though. Currently it's part of a separate job in the unit test workflow which doesn't work because jobs are different vms.

Reply:
@OliverMKing so it is totally okay to run unit tests on push events and separate it from e2e workflows ? Yes unit test and gocover has to run together so that coverage.out can be generated at runtime.

Reply:
We decided to use coverall in par with other OSS repo's like agent baker etc..